tempesta_fw/sock.c

```diff
@@ -1362,18 +1377,34 @@ ss_tx_action(void)
 				bh_unlock_sock(sk);
 				break;
 			}
-			__sk_close_locked(sk); /* paired with bh_lock_sock() */
+			/* paired with bh_lock_sock() */
+			__sk_close_locked(sk);
 			break;
 		case SS_CLOSE:
+			/*
+			 * tcp_fin() sets tcp_set_state(sk, TCP_CLOSE_WAIT) when
+			 * we receive FIN from a peer, next the function calls
+			 * sk->sk_state_change(sk), i.e. our
+			 * ss_tcp_state_change() where we call ss_close() now,
+			 * so our closing work is queued into work queue.
+			 * ss_tx_action() sends all pending data and reaches the
+			 * closing work. Now the socket in TCPF_CLOSE_WAIT state
+			 * and we do call __sk_close_locked(). We didn't need it
+			 * before because with previous ss_linkerror() we
+			 * immediately moved from TCPF_CLOSE_WAIT to next states
+			 * like TCP_CLOSE.
```

Comment:
I'd like this to be a description of a code rather than a description of a change. The line "We didn't need it before because with previous ss_linkerror()" talks about what was code before this particular PR. It will be extremely confusing later.

It would be nice to have this a comment about why those particular ESTABLISHED, SYN_SENT, and CLOSE_WAIT are the states where we want to call `__sk_close_locked()`. Something like:

>We were asked to close the socket. If current state is either ESTABLISHED or SYN_SENT, we initiate the closure process. If for some reason other side sent us FIN, we are at CLOSE_WAIT, so the socket closure is in progress, but still need to cleanup on our side. If we get here while the socket in any other state, resources were either already freed or were never allocated.

I'm still not sure. Also, why SYN_RECV is not in the selected states?

Reply:
SYNC_RECV is not needed, because we select the sockets where we are the server from the kernel upon reaching ESTABLISHED (`ss_tcp_state_change ()`), and we control the sockets where we are the client from the beginning of the life cycle

Reply:
> SYNC_RECV is not needed

Oh, I see. Thanks for the explanation.